crates/matrix-sdk-ui/src/sync_service/mod.rs

```diff
@@ -121,34 +125,47 @@ impl SyncService {
 
         async move {
             let Some(report) = receiver.recv().await else {
-                info!("sync service scheduler: channel has been closed?");
+                info!("internal channel has been closed?");
                 return;
             };
 
+            // If one service failed, make sure to request stopping the other one.
             let (stop_room_list, stop_encryption) = match &report.origin {
                 TerminationOrigin::EncryptionSync => (true, false),
                 TerminationOrigin::RoomList => (false, true),
                 TerminationOrigin::Scheduler => (true, true),
             };
 
-            // Put both tasks in the same state: abort and drop them.
-            if let Some(task) = room_list_task.lock().unwrap().take() {
-                task.abort();
-            }
-            if let Some(task) = encryption_sync_task.lock().unwrap().take() {
-                task.abort();
-            }
+            // Stop both services, and wait for the streams to properly finish: at some
+            // point they'll return `None` and will exit their infinite loops,
+            // and their tasks will gracefully terminate.
 
-            // Put both sliding syncs in the same state: stop both.
             if stop_room_list {
                 if let Err(err) = room_list_service.stop_sync() {
-                    warn!("unable to stop room list after encryption sync failed: {err:#}");
+                    error!("unable to stop room list service: {err:#}");
                 }
             }
-            if stop_encryption {
-                if let Some(encryption_sync) = &encryption_sync {
-                    if let Err(err) = encryption_sync.stop() {
-                        warn!("unable to stop encryption sync after room list failed: {err:#}");
+
+            {
+                let task = room_list_task.lock().unwrap().take();
+                if let Some(task) = task {
+                    if let Err(err) = task.await {
```

Comment:
Can't we wait on both `task.await` (line 152 for `room_list_task` and line 167 for `encryption_sync_task`) at the same time, with a `join_all` or something like that? Maybe the fact that `encryption_sync` is optional makes things more messy. In this case, let's keep things as they are.

Reply:
Both tasks are optional (and the "encryption sync" one is even more so), so we'd need to look at the cases and we can join only if both are set. Doesn't seem worth the hassle to me, especially since they're already moving forward independently of being polled (they've been `spawn()`ed before), so `await`ing one won't block the other from making progress.